File: app/src/main/java/com/androidacademy/msk/exerciseproject/screen/news_list/NewsListPresenter.java
package com.androidacademy.msk.exerciseproject.screen.news_list;

import android.support.annotation.NonNull;

import com.androidacademy.msk.exerciseproject.data.database.NewsConverter;
import com.androidacademy.msk.exerciseproject.data.database.dao.NewsDao;
import com.androidacademy.msk.exerciseproject.data.database.entity.DbNewsItem;
import com.androidacademy.msk.exerciseproject.data.network.api.NYTimesApi;
import com.androidacademy.msk.exerciseproject.model.Section;
import com.androidacademy.msk.exerciseproject.screen.base.BasePresenter;
import com.androidacademy.msk.exerciseproject.utils.EnumUtils;
import com.androidacademy.msk.exerciseproject.utils.NewsDataUtils;
import com.androidacademy.msk.exerciseproject.utils.StringUtils;
import com.arellomobile.mvp.InjectViewState;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

@InjectViewState
public class NewsListPresenter extends BasePresenter<NewsListView> {

    @NonNull
    private final NYTimesApi api;
    @NonNull
    private final NewsDao database;

    @NonNull
    private Section currentSelectedSection = Section.HOME;

    @Inject
    public NewsListPresenter(@NonNull NYTimesApi api, @NonNull NewsDao database) {
        this.api = api;
        this.database = database;
    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();
        getViewState().showEmptyView();
    }

    public void onItemClicked(int id) {
        getViewState().openDetailsScreen(id);
    }

    public void onTryAgainButtonClicked() {
        getNews(currentSelectedSection.toString().toLowerCase());
    }

    public void onFabClicked() {
        getNews(currentSelectedSection.toString().toLowerCase());
    }

    public void onSpinnerItemClicked(@NonNull Section section) {
        if (!section.equals(currentSelectedSection)) {
            currentSelectedSection = section;
            getNews(section.name().toLowerCase());
        }
    }

    public void onSetupSpinnerPosition() {
        List<String> categoryList = EnumUtils.convertEnumValuesToCapitalizedList(Section.values());
        int position = categoryList.indexOf(StringUtils.capitalize(currentSelectedSection.name()));
        getViewState().setCurrentSectionInSpinner(position);

    }

    private void getNews(@NonNull String section) {
        compositeDisposable.add(
                api.getNews(section)
                        .map(newsResponse -> {
                            List<DbNewsItem> dbNewsItems = NewsConverter.toDatabase(
                                    newsResponse.getResults(),
                                    section);
                            database.deleteBySection(section);
                            database.insertAll(dbNewsItems);
                            int[] ids = database.getNewsIdBySection(section);
                            return NewsDataUtils.setIds(dbNewsItems, ids);
                        })
                        .onErrorReturn(throwable -> {
                            List<DbNewsItem> news = database.getNewsBySection(section);
                            if (news.isEmpty()) {
                                return null;
                            }
                            return news;
                        })
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .doOnSubscribe(disposable -> getViewState().showProgressBar())
                        .subscribe(
                                news -> getViewState().showNews(news),
                                throwable -> getViewState().showError())
        );
    }
}

File: app/src/main/java/com/androidacademy/msk/exerciseproject/screen/newslist/NewsListPresenter.kt
package com.androidacademy.msk.exerciseproject.screen.newslist

import NewsDataUtils
import com.androidacademy.msk.exerciseproject.data.database.NewsConverter.toDatabase
import com.androidacademy.msk.exerciseproject.data.database.dao.NewsDao
import com.androidacademy.msk.exerciseproject.data.database.entity.DbNewsItem
import com.androidacademy.msk.exerciseproject.data.network.api.NYTimesApi
import com.androidacademy.msk.exerciseproject.model.Section
import com.androidacademy.msk.exerciseproject.screen.base.BasePresenter
import com.androidacademy.msk.exerciseproject.utils.SectionUtils
import com.arellomobile.mvp.InjectViewState
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject


@InjectViewState
class NewsListPresenter @Inject constructor(
        private val api: NYTimesApi,
        private val newsDao: NewsDao
) : BasePresenter<NewsListView>() {

    private var currentSelectedSection = Section.HOME


    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        viewState.showProgressBar()
        getCurrentSectionNews()
    }


    fun onItemClicked(id: Int) {
        viewState.openDetailsScreen(id)
    }

    fun onTryAgainButtonClicked() {
        viewState.toggleSwipeRefreshProgress(true)
        getCurrentSectionNews()
    }

    fun onSpinnerItemClicked(section: Section) {
        if (section != currentSelectedSection) {
            currentSelectedSection = section

            viewState.toggleSwipeRefreshProgress(true)
            val sectionString = SectionUtils.getSectionForQuery(section)
            getNews(sectionString)
        }
    }


    fun onRefreshSwiped() {
        getCurrentSectionNews()
    }

    private fun getCurrentSectionNews() {
        getNews(currentSelectedSection.toString().toLowerCase())
    }

    private fun getNews(section: String) {
        compositeDisposable.add(
                api.getNews(section)
                        .map<List<DbNewsItem>> { newsResponse ->
                            val dbNewsItems = toDatabase(newsResponse.results, section)
                            newsDao.deleteBySection(section)
                            newsDao.insertAll(dbNewsItems)
                            val ids = newsDao.getNewsIdBySection(section)
                            NewsDataUtils.setIds(dbNewsItems, ids)
                        }
                        .onErrorReturn {
                            val news = newsDao.getNewsBySection(section)
                            if (news.isNotEmpty()) news else null
                        }
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .doFinally { viewState.toggleSwipeRefreshProgress(false) }
                        .subscribe(
                                { news -> viewState.showNews(news) },
                                { viewState.showError() }
                        )
        )
    }
}